ThreadLocalStorage needs a move operator=

For #10284, it turns out that I need a move `operator=` for `AffineConstraints`, but that's not so easy because that class will have a `ThreadLocalStorage` which does not have a move version of `operator=`. (It doesn't have one because it has a `std::mutex` member, which makes the compiler-generated move `operator=` `=delete`d.)

@tamiko -- do you see a good reason not to add one?

I'm confused. The `ThreadLocalStorage` class has a move `operator=` and move constructors alright. And `AffineConstraints` has its move constructor `=default`ed, yet this is the error I get:
```
2303: /raid/bangerth/p/deal.II/1/dealii/tests/bits/periodicity_06.cc:117:10: error: use of deleted function ‘dealii::AffineConstraints<number>::AffineConstraints(dealii::AffineConstraints<number>&&) [with number = double]’
2303:    return constraints;
2303:           ^~~~~~~~~~~
2303: In file included from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_accessor.templates.h:34,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_accessor.h:2085,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_handler.h:33,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/tests/bits/periodicity_06.cc:24:
2303: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/lac/affine_constraints.h:581:3: note: ‘dealii::AffineConstraints<number>::AffineConstraints(dealii::AffineConstraints<number>&&) [with number = double]’ is implicitly deleted because the default definition would be ill-formed:
2303:    AffineConstraints(AffineConstraints &&affine_constraints) = default; // NOLINT
2303:    ^~~~~~~~~~~~~~~~~
2303: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/lac/affine_constraints.h:581:3: error: use of deleted function ‘dealii::Threads::ThreadLocalStorage<T>::ThreadLocalStorage(const dealii::Threads::ThreadLocalStorage<T>&) [with T = dealii::internal::AffineConstraints::ScratchData<double>]’
```
Note how `AffineConstraints(AffineConstraints &&)` is calling `ThreadLocalStorage(const ThreadLocalStorage &)` despite the fact that there is `ThreadLocalStorage(ThreadLocalStorage&&)`.

Similarly:
```
2303: /raid/bangerth/p/deal.II/1/dealii/tests/bits/periodicity_06.cc:252:61: error: use of deleted function ‘dealii::AffineConstraints<number>& dealii::AffineConstraints<number>::operator=(dealii::AffineConstraints<number>&&) [with number = double]’
2303:        constraints[i] = make_constraint_matrix(dof_handler, i);
2303:                                                              ^
2303: In file included from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_accessor.templates.h:34,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_accessor.h:2085,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/include/deal.II/dofs/dof_handler.h:33,
2303:                  from /raid/bangerth/p/deal.II/1/dealii/tests/bits/periodicity_06.cc:24:
2303: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/lac/affine_constraints.h:599:3: note: ‘dealii::AffineConstraints<number>& dealii::AffineConstraints<number>::operator=(dealii::AffineConstraints<number>&&) [with number = double]’ is implicitly deleted because the default definition would be ill-formed:
2303:    operator=(AffineConstraints &&affine_constraints) = default; // NOLINT
2303:    ^~~~~~~~
2303: /raid/bangerth/p/deal.II/1/dealii/include/deal.II/lac/affine_constraints.h:599:3: error: use of deleted function ‘dealii::Threads::ThreadLocalStorage<dealii::internal::AffineConstraints::ScratchData<double> >& dealii::Threads::ThreadLocalStorage<dealii::internal::AffineConstraints::ScratchData<double> >::operator=(const dealii::Threads::ThreadLocalStorage<dealii::internal::AffineConstraints::ScratchData<double> >&)’
```

So are the compiler-generated move-operations supposed to call copy-operations of member variables? Why don't they call move-operations of members?

--- a/include/deal.II/base/thread_local_storage.h
+++ b/include/deal.II/base/thread_local_storage.h
@@ -113,6 +113,17 @@ namespace Threads
     ThreadLocalStorage() = default;
 
     /**
+     * Copy constructor.
+     */
+    ThreadLocalStorage(const ThreadLocalStorage &);
+
+    /**
+     * Move constructor. The constructor moves all internal data structures
+     * from the argument.
+     */
+    ThreadLocalStorage(ThreadLocalStorage &&);
+
+    /**
      * A kind of copy constructor. Initializes an internal exemplar by the
      * given object. The exemplar is in turn used to initialize each thread
      * local object instead of invoking the default constructor.
@@ -127,10 +138,16 @@ namespace Threads
     explicit ThreadLocalStorage(T &&t);
 
     /**
-     * The copy constructor is deleted. Copying instances of this class is not
-     * allowed.
+     * Copy assignment operator.
      */
-    ThreadLocalStorage(const ThreadLocalStorage<T> &t) = delete;
+    ThreadLocalStorage &
+    operator=(const ThreadLocalStorage &t);
+
+    /**
+     * Move assignment operator.
+     */
+    ThreadLocalStorage &
+    operator=(ThreadLocalStorage &&t);
 
     /**
      * Return a reference to the data stored by this object for the current
@@ -253,6 +270,32 @@ namespace Threads
 {
   // ----------------- inline and template functions --------------------------
 
+
+  template <typename T>
+  ThreadLocalStorage<T>::ThreadLocalStorage(const ThreadLocalStorage<T> &t)
+    : exemplar(t.exemplar)
+  {
+    /*
+     * Raise a reader lock while we are populating our own data in order to
+     * avoid copying over an invalid state.
+     */
+    std::shared_lock<decltype(insertion_mutex)> lock(t.insertion_mutex);
+    data = t.data;
+  }
+
+
+  template <typename T>
+  ThreadLocalStorage<T>::ThreadLocalStorage(ThreadLocalStorage<T> &&t)
+    : exemplar(t.exemplar)
+  {
+    /*
+     * We are nice and raise the writer lock before copying over internal
+     * data structures from the argument.
+     */
+    std::unique_lock<decltype(insertion_mutex)> lock(t.insertion_mutex);
+    data = std::move(t.data);
+  }
+
   template <typename T>
   inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
     : exemplar(std::make_shared<const T>(t))
@@ -263,6 +306,36 @@ namespace Threads
   inline ThreadLocalStorage<T>::ThreadLocalStorage(T &&t)
     : exemplar(std::make_shared<T>(std::forward<T>(t)))
   {}
+
+
+  template <typename T>
+  inline ThreadLocalStorage<T> &
+  ThreadLocalStorage<T>::operator=(const ThreadLocalStorage<T> &t)
+  {
+    /*
+     * We need to raise the reader lock of the argument and our writer lock
+     * while copying internal data structures.
+     */
+    std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
+    std::unique_lock<decltype(insertion_mutex)> writer_lock(insertion_mutex);
+
+    data = t.data;
+  }
+
+
+  template <typename T>
+  inline ThreadLocalStorage<T> &
+  ThreadLocalStorage<T>::operator=(ThreadLocalStorage<T> &&t)
+  {
+    /*
+     * We need to raise the reader lock of the argument and our writer lock
+     * while copying internal data structures.
+     */
+    std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
+    std::unique_lock<decltype(insertion_mutex)> writer_lock(insertion_mutex);
+
+    data = std::move(t.data);
+  }
 
 
 #    ifndef DOXYGEN

--- a/include/deal.II/base/thread_local_storage.h
+++ b/include/deal.II/base/thread_local_storage.h
@@ -285,7 +285,7 @@ namespace Threads
 
   template <typename T>
   ThreadLocalStorage<T>::ThreadLocalStorage(ThreadLocalStorage<T> &&t) noexcept
-    : exemplar(t.exemplar)
+    : exemplar(std::move(t.exemplar))
   {
     // We are nice and raise the writer lock before copying over internal
     // data structures from the argument.

--- a/include/deal.II/base/thread_local_storage.h
+++ b/include/deal.II/base/thread_local_storage.h
@@ -121,7 +121,7 @@ namespace Threads
      * Move constructor. The constructor moves all internal data structures
      * from the argument.
      */
-    ThreadLocalStorage(ThreadLocalStorage &&);
+    ThreadLocalStorage(ThreadLocalStorage &&t) noexcept;
 
     /**
      * A kind of copy constructor. Initializes an internal exemplar by the
@@ -147,7 +147,7 @@ namespace Threads
      * Move assignment operator.
      */
     ThreadLocalStorage &
-    operator=(ThreadLocalStorage &&t);
+    operator=(ThreadLocalStorage &&t) noexcept;
 
     /**
      * Return a reference to the data stored by this object for the current
@@ -275,31 +275,47 @@ namespace Threads
   ThreadLocalStorage<T>::ThreadLocalStorage(const ThreadLocalStorage<T> &t)
     : exemplar(t.exemplar)
   {
-    /*
-     * Raise a reader lock while we are populating our own data in order to
-     * avoid copying over an invalid state.
-     */
+    // Raise a reader lock while we are populating our own data in order to
+    // avoid copying over an invalid state.
     std::shared_lock<decltype(insertion_mutex)> lock(t.insertion_mutex);
     data = t.data;
   }
 
 
+
   template <typename T>
-  ThreadLocalStorage<T>::ThreadLocalStorage(ThreadLocalStorage<T> &&t)
+  ThreadLocalStorage<T>::ThreadLocalStorage(ThreadLocalStorage<T> &&t) noexcept
     : exemplar(t.exemplar)
   {
-    /*
-     * We are nice and raise the writer lock before copying over internal
-     * data structures from the argument.
-     */
+    // We are nice and raise the writer lock before copying over internal
+    // data structures from the argument.
+    //
+    // The point is a bit moot, though: Users of ThreadLocalStorage
+    // typically obtain their thread's thread-local object through the
+    // get() function. That function also acquires the lock, but
+    // whether or not we do that here really doesn't make any
+    // difference in terms of correctness: If another thread manages
+    // to call get() just before we get here, then the result of that
+    // get() function immediately becomes invalid; if it manages to
+    // call get() at the same time as this function if there were no
+    // locking here, it might access undefined state; and if it
+    // manages to call get() just after we moved away the state --
+    // well, then it just got lucky to escape the race condition, but
+    // the race condition is still there.
+    //
+    // On the other hand, there is no harm in doing at least
+    // conceptually the right thing, so ask for that lock:
     std::unique_lock<decltype(insertion_mutex)> lock(t.insertion_mutex);
     data = std::move(t.data);
   }
+
+
 
   template <typename T>
   inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
     : exemplar(std::make_shared<const T>(t))
   {}
+
 
 
   template <typename T>
@@ -308,14 +324,13 @@ namespace Threads
   {}
 
 
+
   template <typename T>
   inline ThreadLocalStorage<T> &
   ThreadLocalStorage<T>::operator=(const ThreadLocalStorage<T> &t)
   {
-    /*
-     * We need to raise the reader lock of the argument and our writer lock
-     * while copying internal data structures.
-     */
+    // We need to raise the reader lock of the argument and our writer lock
+    // while copying internal data structures.
     std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
     std::unique_lock<decltype(insertion_mutex)> writer_lock(insertion_mutex);
 
@@ -323,15 +338,18 @@ namespace Threads
   }
 
 
+
   template <typename T>
   inline ThreadLocalStorage<T> &
-  ThreadLocalStorage<T>::operator=(ThreadLocalStorage<T> &&t)
+  ThreadLocalStorage<T>::operator=(ThreadLocalStorage<T> &&t) noexcept
   {
-    /*
-     * We need to raise the reader lock of the argument and our writer lock
-     * while copying internal data structures.
-     */
-    std::shared_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
+    // We need to raise the writer lock of the argument (because we're
+    // moving information *away* from that object) and the writer lock
+    // of our object while copying internal data structures.
+    //
+    // That said, the same issue with acquiring the source lock as
+    // with the move constructor above applies here as well.
+    std::unique_lock<decltype(insertion_mutex)> reader_lock(t.insertion_mutex);
     std::unique_lock<decltype(insertion_mutex)> writer_lock(insertion_mutex);
 
     data = std::move(t.data);

--- a/include/deal.II/lac/affine_constraints.h
+++ b/include/deal.II/lac/affine_constraints.h
@@ -23,6 +23,7 @@
 #include <deal.II/base/subscriptor.h>
 #include <deal.II/base/table.h>
 #include <deal.II/base/template_constraints.h>
+#include <deal.II/base/thread_local_storage.h>
 
 #include <deal.II/lac/vector.h>
 #include <deal.II/lac/vector_element_access.h>

--- a/include/deal.II/lac/affine_constraints.h
+++ b/include/deal.II/lac/affine_constraints.h
@@ -1830,6 +1830,10 @@ private:
    */
   bool sorted;
 
+  mutable Threads::ThreadLocalStorage<
+    internal::AffineConstraints::ScratchData<number>>
+    scratch_data;
+
   /**
    * Internal function to calculate the index of line @p line_n in the vector
    * lines_cache using local_lines.

--- a/include/deal.II/lac/affine_constraints.templates.h
+++ b/include/deal.II/lac/affine_constraints.templates.h
@@ -2796,11 +2796,6 @@ namespace internal
       private:
         ScratchData<number> *my_scratch_data;
       };
-
-      /**
-       * The actual data object that contains a scratch data for each thread.
-       */
-      static Threads::ThreadLocalStorage<ScratchData<number>> scratch_data;
     };
 
 
@@ -3560,9 +3555,8 @@ AffineConstraints<number>::distribute_local_to_global(
 
   const size_type n_local_dofs = local_dof_indices.size();
 
-  typename internal::AffineConstraints::AffineConstraintsData<number>::
-    ScratchDataAccessor scratch_data(
-      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
+  typename internal::AffineConstraints::AffineConstraintsData<
+    number>::ScratchDataAccessor scratch_data(this->scratch_data);
 
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
     scratch_data->global_rows;
@@ -3712,9 +3706,8 @@ AffineConstraints<number>::distribute_local_to_global(
     }
   Assert(sorted == true, ExcMatrixNotClosed());
 
-  typename internal::AffineConstraints::AffineConstraintsData<number>::
-    ScratchDataAccessor scratch_data(
-      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
+  typename internal::AffineConstraints::AffineConstraintsData<
+    number>::ScratchDataAccessor scratch_data(this->scratch_data);
 
   const size_type n_local_dofs = local_dof_indices.size();
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
@@ -3851,8 +3844,7 @@ AffineConstraints<number>::distribute_local_to_global(
 
   typename internal::AffineConstraints::AffineConstraintsData<
     typename MatrixType::value_type>::ScratchDataAccessor
-    scratch_data(
-      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
+    scratch_data(this->scratch_data);
 
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
     scratch_data->global_rows;
@@ -3911,9 +3903,8 @@ AffineConstraints<number>::add_entries_local_to_global(
          ExcNotQuadratic());
 
   const size_type n_local_dofs = local_dof_indices.size();
-  typename internal::AffineConstraints::AffineConstraintsData<number>::
-    ScratchDataAccessor scratch_data(
-      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
+  typename internal::AffineConstraints::AffineConstraintsData<
+    number>::ScratchDataAccessor scratch_data(this->scratch_data);
 
   const bool dof_mask_is_active = (dof_mask.n_rows() == n_local_dofs);
   if (dof_mask_is_active == true)
@@ -4068,9 +4059,8 @@ AffineConstraints<number>::add_entries_local_to_global(
   const size_type n_local_dofs = local_dof_indices.size();
   const size_type num_blocks   = sparsity_pattern.n_block_rows();
 
-  typename internal::AffineConstraints::AffineConstraintsData<number>::
-    ScratchDataAccessor scratch_data(
-      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
+  typename internal::AffineConstraints::AffineConstraintsData<
+    number>::ScratchDataAccessor scratch_data(this->scratch_data);
 
   const bool dof_mask_is_active = (dof_mask.n_rows() == n_local_dofs);
   if (dof_mask_is_active == true)

--- a/include/deal.II/lac/affine_constraints.templates.h
+++ b/include/deal.II/lac/affine_constraints.templates.h
@@ -2754,11 +2754,12 @@ namespace internal
       {
       public:
         /**
-         * Constructor. Grabs the scratch data object on the current thread and
-         * marks it as used.
+         * Constructor. Takes the scratch data object for the current
+         * thread out of the provided object and marks it as used.
          */
-        ScratchDataAccessor()
-          : my_scratch_data(&AffineConstraintsData::scratch_data.get())
+        ScratchDataAccessor(
+          Threads::ThreadLocalStorage<ScratchData<number>> &tls_scratch_data)
+          : my_scratch_data(&tls_scratch_data.get())
         {
           Assert(
             my_scratch_data->in_use == false,
@@ -2796,7 +2797,6 @@ namespace internal
         ScratchData<number> *my_scratch_data;
       };
 
-    private:
       /**
        * The actual data object that contains a scratch data for each thread.
        */
@@ -3560,8 +3560,9 @@ AffineConstraints<number>::distribute_local_to_global(
 
   const size_type n_local_dofs = local_dof_indices.size();
 
-  typename internal::AffineConstraints::AffineConstraintsData<
-    number>::ScratchDataAccessor scratch_data;
+  typename internal::AffineConstraints::AffineConstraintsData<number>::
+    ScratchDataAccessor scratch_data(
+      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
 
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
     scratch_data->global_rows;
@@ -3711,8 +3712,9 @@ AffineConstraints<number>::distribute_local_to_global(
     }
   Assert(sorted == true, ExcMatrixNotClosed());
 
-  typename internal::AffineConstraints::AffineConstraintsData<
-    number>::ScratchDataAccessor scratch_data;
+  typename internal::AffineConstraints::AffineConstraintsData<number>::
+    ScratchDataAccessor scratch_data(
+      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
 
   const size_type n_local_dofs = local_dof_indices.size();
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
@@ -3848,7 +3850,9 @@ AffineConstraints<number>::distribute_local_to_global(
   const size_type n_local_col_dofs = col_indices.size();
 
   typename internal::AffineConstraints::AffineConstraintsData<
-    typename MatrixType::value_type>::ScratchDataAccessor scratch_data;
+    typename MatrixType::value_type>::ScratchDataAccessor
+    scratch_data(
+      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
 
   internal::AffineConstraints::GlobalRowsFromLocal<number> &global_rows =
     scratch_data->global_rows;
@@ -3907,8 +3911,9 @@ AffineConstraints<number>::add_entries_local_to_global(
          ExcNotQuadratic());
 
   const size_type n_local_dofs = local_dof_indices.size();
-  typename internal::AffineConstraints::AffineConstraintsData<
-    number>::ScratchDataAccessor scratch_data;
+  typename internal::AffineConstraints::AffineConstraintsData<number>::
+    ScratchDataAccessor scratch_data(
+      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
 
   const bool dof_mask_is_active = (dof_mask.n_rows() == n_local_dofs);
   if (dof_mask_is_active == true)
@@ -4063,8 +4068,9 @@ AffineConstraints<number>::add_entries_local_to_global(
   const size_type n_local_dofs = local_dof_indices.size();
   const size_type num_blocks   = sparsity_pattern.n_block_rows();
 
-  typename internal::AffineConstraints::AffineConstraintsData<
-    number>::ScratchDataAccessor scratch_data;
+  typename internal::AffineConstraints::AffineConstraintsData<number>::
+    ScratchDataAccessor scratch_data(
+      internal::AffineConstraints::AffineConstraintsData<number>::scratch_data);
 
   const bool dof_mask_is_active = (dof_mask.n_rows() == n_local_dofs);
   if (dof_mask_is_active == true)

--- a/source/lac/affine_constraints.cc
+++ b/source/lac/affine_constraints.cc
@@ -137,35 +137,4 @@ dealii::AffineConstraints<double>::distribute<
 #  endif
 #endif
 
-/*
- * Allocate scratch data.
- *
- * We cannot use the generic template instantiation because we need to
- * provide an initializer object of type
- * internals::AffineConstraintsData<Number> that can be passed to the
- * constructor of scratch_data (it won't allow one to be constructed in
- * place).
- */
-
-namespace internal
-{
-  namespace AffineConstraints
-  {
-#define SCRATCH_INITIALIZER(number, Name)              \
-  ScratchData<number> scratch_data_initializer_##Name; \
-  template <>                                          \
-  Threads::ThreadLocalStorage<ScratchData<number>>     \
-    AffineConstraintsData<number>::scratch_data(       \
-      scratch_data_initializer_##Name)
-
-    SCRATCH_INITIALIZER(double, d);
-    SCRATCH_INITIALIZER(float, f);
-#ifdef DEAL_II_WITH_COMPLEX_VALUES
-    SCRATCH_INITIALIZER(std::complex<double>, cd);
-    SCRATCH_INITIALIZER(std::complex<float>, cf);
-#endif
-#undef SCRATCH_INITIALIZER
-  } // namespace AffineConstraints
-} // namespace internal
-
 DEAL_II_NAMESPACE_CLOSE

